lib/Sema/TypeCheckStmt.cpp

```diff
@@ -508,7 +508,8 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
       Type contextType = yieldType;
       if (yieldResults[i].isInOut()) {
         contextTypePurpose = CTP_YieldByReference;
-        contextType = LValueType::get(contextType);
+        if (!contextType->hasError())
+          contextType = LValueType::get(contextType);
```

Comment:
@marcrasi @rjmccall I think like other statements this one should return `nullptr` if at least one of the result types is null or error. Better yet it should also produce a diagnostic somewhere so we don't get into AST verifier.

Reply:
Returning `nullptr` does sound better. I have updated the PR to do that.

I don't think it would be right to produce a diagnostic, because whatever produced the error type should have already produced a diagnostic?

Diagnostics do not prevent the ASTVerifier from running during code-completion, so having diagnostics doesn't fix the ASTVerifier problem. Perhaps diagnostics should stop the ASTVerifier from running? Happy to try to do that, if it's the right thing to do.

The specific ASTVerifier error triggered by this testcase is:
```
checked decl cannot have error type
(var_decl range=[<REPL Input>:8:18 - line:8:18] "ba" type='<<error type>>' interface type='<<error type>>' access=internal readImpl=stored writeImpl=stored readWriteImpl=stored
  (accessor_decl implicit range=[<REPL Input>:8:18 - line:8:18] 'anonname=0x83096e8' interface type='(Foo) -> () -> <<error type>>' access=internal get_for=ba
    (parameter "self" interface type='Foo')
    (parameter_list))
  (accessor_decl implicit range=[<REPL Input>:8:18 - line:8:18] 'anonname=0x83098b8' interface type='(inout Foo) -> (<<error type>>) -> ()' access=internal set_for=ba
    (parameter "self" interface type='Foo' inout)
    (parameter_list
      (parameter "value" interface type='<<error type>>') range=[<REPL Input>:8:18 - line:8:18]))
  (accessor_decl implicit range=[<REPL Input>:8:18 - line:8:18] 'anonname=0x83099b0' interface type='(inout Foo) -> () -> ()' access=internal _modify_for=ba
    (parameter "self" interface type='Foo' inout)
    (parameter_list)
    (brace_stmt implicit range=[<REPL Input>:8:18 - line:8:18]
      (yield_stmt implicit range=[<REPL Input>:8:18 - line:8:18]
        (inout_expr implicit type='inout <<error type>>'
          (member_ref_expr implicit type='<null>' decl=swift_ide_test.(file).Foo.ba@<REPL Input>:8:18 direct_to_impl
            (declref_expr implicit type='<null>' decl=swift_ide_test.(file).Foo.<anonymous>.self@<REPL Input>:8:18 function_ref=unapplied)))))))
Stack dump:
0.	Program arguments: /usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test -target x86_64-unknown-linux-gnu -module-cache-path /usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/swift-test-results/x86_64-unknown-linux-gnu/clang-module-cache -completion-cache-path /usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/swift-test-results/x86_64-unknown-linux-gnu/completion-cache -swift-version 4 -repl-code-completion -source-filename=/usr/local/google/home/marcrasi/swift-base-master/swift/test/IDE/complete_tf_195.swift 
1.	While walking into decl 'Foo' (at <REPL Input>:8:1)
2.	While verifying VarDecl 'ba' (at <REPL Input>:8:18)
#0 0x0000000003f5bcd4 PrintStackTraceSignalHandler(void*) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x3f5bcd4)
#1 0x0000000003f59e30 llvm::sys::RunSignalHandlers() (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x3f59e30)
#2 0x0000000003f5be82 SignalHandler(int) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x3f5be82)
#3 0x00007eff3b33a0c0 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x110c0)
#4 0x00007eff39702fcf gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x32fcf)
#5 0x00007eff397043fa abort (/lib/x86_64-linux-gnu/libc.so.6+0x343fa)
#6 0x00000000010b6236 (anonymous namespace)::Verifier::verifyChecked(swift::ValueDecl*) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x10b6236)
#7 0x00000000010b67da (anonymous namespace)::Verifier::verifyChecked(swift::VarDecl*) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x10b67da)
#8 0x00000000010ac8cb (anonymous namespace)::Verifier::walkToDeclPost(swift::Decl*) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x10ac8cb)
#9 0x00000000010b8138 (anonymous namespace)::Traversal::doIt(swift::Decl*) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x10b8138)
#10 0x00000000010bcc16 swift::ASTVisitor<(anonymous namespace)::Traversal, swift::Expr*, swift::Stmt*, bool, swift::Pattern*, bool, void>::visit(swift::Pattern*) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x10bcc16)
#11 0x00000000010bdba7 swift::ASTVisitor<(anonymous namespace)::Traversal, swift::Expr*, swift::Stmt*, bool, swift::Pattern*, bool, void>::visit(swift::Decl*) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x10bdba7)
#12 0x00000000010b811c (anonymous namespace)::Traversal::doIt(swift::Decl*) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x10b811c)
#13 0x00000000010be12b (anonymous namespace)::Traversal::visitNominalTypeDecl(swift::NominalTypeDecl*) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x10be12b)
#14 0x00000000010b811c (anonymous namespace)::Traversal::doIt(swift::Decl*) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x10b811c)
#15 0x00000000010b8043 swift::Decl::walk(swift::ASTWalker&) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x10b8043)
#16 0x00000000011657ee swift::SourceFile::walk(swift::ASTWalker&) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x11657ee)
#17 0x00000000010a2e63 swift::verify(swift::SourceFile&) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x10a2e63)
#18 0x0000000000d4dd40 swift::performTypeChecking(swift::SourceFile&, swift::TopLevelContext&, swift::OptionSet<swift::TypeCheckingFlags, unsigned int>, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0xd4dd40)
#19 0x000000000058beda doCodeCompletion(swift::SourceFile&, llvm::StringRef, unsigned int*, swift::CodeCompletionCallbacksFactory*) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x58beda)
#20 0x000000000058b960 swift::REPLCompletions::populate(swift::SourceFile&, llvm::StringRef) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x58b960)
#21 0x0000000000496bff main (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x496bff)
#22 0x00007eff396f02b1 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x202b1)
#23 0x000000000049148a _start (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x49148a)
/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/test-linux-x86_64/IDE/Output/complete_tf_195.swift.script: line 1: 202423 Aborted                 /usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test -target x86_64-unknown-linux-gnu -module-cache-path '/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/swift-test-results/x86_64-unknown-linux-gnu/clang-module-cache' -completion-cache-path '/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/swift-test-results/x86_64-unknown-linux-gnu/completion-cache' -swift-version 4 -repl-code-completion -source-filename=/usr/local/google/home/marcrasi/swift-base-master/swift/test/IDE/complete_tf_195.swift
```

Reply:
> I don't think it would be right to produce a diagnostic, because whatever produced the error type should have already produced a diagnostic?

That's what I meant, code which produces error type should also produce a diagnostic.

> Diagnostics do not prevent the ASTVerifier from running during code-completion, so having diagnostics doesn't fix the ASTVerifier problem.

Error diagnostic should stop verification of type-checked AST - https://github.com/apple/swift/blob/master/lib/AST/ASTVerifier.cpp#L465

Reply:
As far as I remember code completion would avoid displaying produced diagnostics because they have a special sync consumer for them. /cc @rintaro

Reply:
Ah! I think what is happening is that REPLCodeCompletion.cpp is [creating a DiagnosticTransaction](https://github.com/apple/swift/blob/706f15f228b20daf901a3d666f68dd0f8826dbc7/lib/IDE/REPLCodeCompletion.cpp#L196) to prevent diagnostics from printing out, but unfortunately that also prevents the diagnostics from getting put in the ASTContext, which makes stuff like the ASTVerifier incorrectly think that there are no errors when they call ASTContext.hadError().

Maybe I could change REPLCodeCompletion to suppress diagnostic printing in a different way. Like set some flag on the DiagnosticEngine that tells it to collect diagnostics like normal but never actually print them out. This way all the passes and verifiers will correctly see if there were errors, but the errors will not get printed out to the console.

Reply:
Ah, @marcrasi could you try [DiagnosticSuppression](https://github.com/apple/swift/blob/706f15f228b20daf901a3d666f68dd0f8826dbc7/include/swift/AST/DiagnosticSuppression.h#L30) instead?

Reply:
That seems to work! I created a new PR for that: https://github.com/apple/swift/pull/22591

Once that PR merges, the test in this PR should work with the ASTVerifier enabled. I will update this PR once that PR merges.

Reply:
Actually, #22591 completely fixes this crash. Probably because this crash happens in a later phase of typechecking that gets skipped now that #22591 makes it possible to see that there were errors.

So I will close this PR.